feat(lua): vim.region() visual selection (FKA get_visual_selection)

anihm136 opened the pull request:
Function returns the current/last visual selection as a string. Potential changes/considerations that may come under the scope of this PR -
- Refactor the function to separate getting the region demarcated by two marks, and converting the region to a string
- Allow specifying marks to get the region between any two marks (instead of just the visual region marks)
- Reuse region selection in highlight.vim for yanked region highlighting
- Returning a table instead of a string might be helpful
- Removing indentation from each line might be helpful (at least as an option)

**Known issues**
- ~~When the last character in the selection is a multibyte unicode character, the last character is not copied correctly. This appears to be an issue with `vim.region`, where only one byte is copied of the last character~~
- When a visual selection is active, the contents of the last visual selection are returned (instead of the currently active visual selection)
- Does not work as a command (`:lua vim.get_visual_selection()`), and does not work in visual block mode

Pinging @clason, @bfredl, @tjdevries involved in the discussion on Gitter/Matrix

anihm136 committed:
vim.lua: Add vim.get_visual_selection()

Returns the contents of the current/last visual selection as a string

bfredl commented:
Maybe `strpart()` (with len=1 and chars=trua) can be used?

anihm136 committed:
vim.lua: Use strpart in vim.get_visual_selection()

Replace lua string substitution with strpart(), fixing
the issue when the last character of the selection is a
multibyte unicode character

anihm136 commented:
Thanks, `strpart()` fixes the issue with multibyte unicode chars! I had to do some wrangling of the start and end columns, but it works fine now. 
However I see an important issue - the function returns the _last_ visual selection instead of the current one when there is currently a visual selection. This happens because the `'<` and `'>` marks are updated after the visual mode ends. Need to find a solution for that.
A slightly smaller issue, the function cannot be called correctly as a command (at least right now) as `:lua print(vim.get_visual_selection())`, because it changes to commandline mode when you do it that way. It would be nice to find a way to detect that a visual selection was active just before entering commandline mode (I'm not sure if this would be very useful, but nice to have anyway)

clason commented:
> Thanks, `strpart()` fixes the issue with multibyte unicode chars! I had to do some wrangling of the start and end columns, but it works fine now.

Nice! Could you try if that fix also applies to `vim.region`/`highlight_region`, which also has problems with multibyte characters?

tjdevries commented:
Have you tried calling it with a <cmd> mapping instead of :

?

That might fix it?

anihm136 commented:
@clason What is the exact issue with `highlight_region`? I can take a look at it and see if `strpart` can be used to fix it. I suppose it would not belong to this PR though?

@tjdevries Using a mapping works partially i.e, it detects the visual mode correctly (and therefore runs), however the `'<` and `'>` marks are not set until after leaving visual mode. So for instance, I can either call it as a command and it will detect the 'current' selection (which is actually the last selection, since the user left visual mode), but future calls will still return the last selection even if there is no current visual selection. Or I can call it through a mapping, which correctly detects when there is an active visual selection and only runs then, but does not detect the current visual selection and uses the previous one instead. (I hope that was clear, it's kinda hard to explain for me :))

tjdevries commented:
One thing as well I was thinking is I want to have a function that takes marks and gets the region of text.

Then this visual function can call it.

Vim.marsk_to_region or something like that.

Idk I'm on mobile and about to head to church 😂 I'll respond more later.

clason commented:
> One thing as well I was thinking is I want to have a function that takes marks and gets the region of text.
> 
> Then this visual function can call it.

Yes, that is what I meant on gitter by "refactoring".

> @clason What is the exact issue with highlight_region? I can take a look at it and see if strpart can be used to fix it. I suppose it would not belong to this PR though?

You can see it by enabling `highlight_onyank` and yanking a block visual selection involving multibyte characters (kanji, emoji, etc.) You'll see that the highlights do not correspond to the yanked selection but are off due to wrong counting of the characters. I'd have to check again, but if I remember correctly, `vim.region` already reports the wrong columns in this case.

fdschmidt93 commented:
@anihm136 

If possible, try vim.fn.getpos('v') (start) and vim.fn.getcurpos() (quasi-end of visual mode) which are updated live as opposed to '< and '> registers that are only updated after leaving visual mode, as you've said. Though they are not strictly registers.

I am working on the same thing for telescope and was trying to fix visual selection for non-standard ASCII characters there next, so cool to see this progress :)

clason commented:
> Visual block mode likely cannot be fully supported as there are very difficult edge cases with multi-width chars, especially when resolving beginning of selection across lines with varying char widths. One workaround would be to copy selection into an [intermediate register](https://github.com/jpalardy/vim-slime/blob/72171eaaf176d7eb3f73ecebf86ff3e5f4ba7dbd/autoload/slime.vim#L315) like Vim-slime.

For the record, this is exactly what I got stumped on with `vim.region`.

anihm136 committed:
vim.lua: Refactor vim.get_visual_selection()

Separate functionality into vim.get_marked_region() and
vim.get_visual_selection(). Use 'v' and '.' for getting visual selection region,
and adjust positions based on visual mode

anihm136 committed:
vim.lua: Fix minor errors

anihm136 commented:
Okay, I have refactored the function into separate functions for getting the region between two marks and then using that to get the visual selection. @tjdevries can you take a look at `vim.get_marked_region` and let me know if that was what you were thinking about?
Also I spoke too soon about strpart fixing the issue, because `vim.region` returns a byte index, not a char index. @fdschmidt93 I saw your implementation and I think your function for resolving the column recursively is just what is needed here to get it working with multibyte characters. Not sure if it's best to add it to this PR or make a new one to add multibyte character support.
Apart from that, it is working in general for both visual char and visual line modes, both directions of selection

fdschmidt93 commented:
@anihm136 

I'm also very close to full v-block mode support - hopefully only multi-width backwards v-block edge cases to iron out. Once finished, I'll commit it to telescope for review to ultimately make a PR to refactor `vim.region` with full char, line, and block-selection support. (telescope commit will ease review and debugging as it returns visual selection rather than region)

In addition, `vim.get_visual_selection` probably should feature arguments `trim` (default=true) and `delimiter` (default=' ' ?) such that users can optionally augment their output. I'm also afraid visual block mode support will require significant changes to how positions of marks are retrieved (will be factored out) as more info from `vim.fn.getpos` is needed than presented here to later resolve visual columns (as in, actual screen columns).

anihm136 commented:
That's great! I'll hold this until you have refactored `vim.region`, and then we can see what changes are to be made to get the functionality we want

clason commented:
@fdschmidt93  https://github.com/neovim/neovim/pull/13978 might make handling multibyte character much easier once merged

glyh commented:
It's been nearly a year, when will this be merged?

clason commented:
When it's done. Remember that everyone here is doing this work in their spare time, which might be very little these days. Such comments do _not_ speed up development (are more likely than not, discourage people).

gpanders commented on src/nvim/lua/vim.lua:
```diff
@@ -457,6 +457,93 @@ function vim.region(bufnr, pos1, pos2, regtype, inclusive)
+--@return region region between the two marks, as returned by |vim.region|
+--@return start (row,col) tuple denoting the start of the region
+--@return finish (row,col) tuple denoting the end of the region
+function vim.get_marked_region(mark1, mark2, options)
```
This can use anything that can be passed to `getpos()`, which includes but is not limited to marks. Calling it `get_marked_region` is misleading and (IMO) a bit confusing.

Can we modify `vim.region` to accept the special characters that `line()` and friends take, in addition to `(row, col)` tuples:

```
.            the cursor position
$            the last line in the current buffer
'x           position of mark x (if the mark is not set, 0 is
             returned)
w0           first line visible in current window (one if the
             display isn't updated, e.g. in silent Ex mode)
w$           last line visible in current window (this is one
             less than "w0" if no lines are visible)
v            In Visual mode: the start of the Visual area (the
             cursor is the end).  When not in Visual mode
             returns the cursor position.  Differs from |'<| in
             that it's updated right away.
```

If so, then we wouldn't need this new function at all and we can just use `vim.region()`.

clason replied:
Should be a straightforward refactor by moving the code (specific to `'[` and `']`) from `vim.highlight.on_yank` up to `vim.region` and just passing those marks through.

(Handling multibyte characters is a separate and more general issue that I'd recommend tackling as a follow-up PR.)

justinmk replied:
Done in https://github.com/neovim/neovim/pull/22635

justinmk reviewed (changes requested):
I think this could be an enhancement/overload to `vim.region` rather than extra functions.

See @gpanders comment: https://github.com/neovim/neovim/pull/13896/files#r779819674

gpanders commented:
@anihm136 You are of course free to jump back in and push this forward when and if you want to, but in the meantime if another contributor is willing to push this PR across the finish line I'm marking this PR as "open to adoption". If you'd like to take over this PR, just drop a line in the [Neovim dev channel](https://matrix.to/#/%23neovim-dev:matrix.org) on Matrix and let the maintainers know you're interested.

Liverm0r commented:
sorry, I can't get how I could get a visual selection :) please, help

glyh commented:
@Liverm0r a snippet that might help you:
```lua
function _M.get_selection()
  -- does not handle rectangular selection
  local s_start = vim.fn.getpos("'<")
  local s_end = vim.fn.getpos("'>")
  local n_lines = math.abs(s_end[2] - s_start[2]) + 1
  local lines = vim.api.buf_get_lines(0, s_start[2] - 1, s_end[2], false)
  lines[1] = string.sub(lines[1], s_start[3], -1)
  if n_lines == 1 then
    lines[n_lines] = string.sub(lines[n_lines], 1, s_end[3] - s_start[3] + 1)
  else
    lines[n_lines] = string.sub(lines[n_lines], 1, s_end[3])
  end
  return table.concat(lines, '\n')
end
```

But I think it would be better if you ask it in neovim's chat servers, reddit, SOF, etc.

hinell commented:
Any progress on this?

Andrew15-5 commented:
I'm currently having trouble with multibyte chars. Basically, I need to create a `get_selection()`. It's not too complicated to do a one line selection, I think. But multiline — I'm not sure whether I can do this (I'm not too great with Vimscript).

And here I am witnessing the creation of the very thing I was looking for, but in Lua (I guess the multibyte part is not an intended purpose). Can someone tell me what seems to be the problem? 2 years have passed.

clason commented:
> Can someone tell me what seems to be the problem? 2 years have passed.

Unicode is _hard_; the multibyte thing is very much an intended purpose. Lots of corner cases.

fdschmidt93 commented:
You might want to have a look at https://github.com/nvim-telescope/telescope.nvim/pull/2333 and adopt what vim-slime is doing.

The implementation of the linked PR should work well for visual mode (not visual line or visual block mode, but would be easily extended to it -- see vim-slime implementation, for reference).

tl;dr: Yank the visual selection into a temporary register, get text from said register, restore register.

clason commented:
Visual block mode is exactly what is blocking this... (Pun intended.)

Andrew15-5 commented:
> tl;dr: Yank the visual selection into a temporary register, get text from said register, restore original register.

The most frustrating thing is that I just tried it a few minutes ago, and it doesn't work! If I execute code line by line in command mode, then it suddenly works as intended but not via function. I literally have no idea why.

```vim
function! Get_selected_text()
  call feedkeys('gvy')
  return @"
endfunction
```

- open Neovim
- write something
- select it
- run `echo Get_selected_text()`
- doesn't work
- run again
- it works

It captures only the **previous** state of `@"`, **not the current** one. This is one of those moments when you (me) think that the program is dumb and not you (but a good program is never dumb, so it must be you).

But doing it manually works:

- open Neovim
- write something
- select it
- run `call feedkeys('gvy')`
- run `echo @"`
- first try, baby

P.S. If I remember correctly, I had the exact same problem really long time ago. Probably why I still don't have this "workaround" function in my arsenal.

fdschmidt93 commented:
We trailing off the path (= PR) a bit here.

I guess your problem occurs, because you are calling the function from within visual mode. In turn, it works, once you exit visual mode ("recalling previous state") and recall (`gv`) the prior visual selection. You can always programmatically check the current mode and return to normal mode if in exit mode. Without looking at the code in telescope, I'm pretty sure that's what's happening there.

Andrew15-5 commented:
> Visual block mode is exactly what is blocking this... (Pun intended.)

Does the _block_ mode _block_ as in no progress is being done (unsolvable edge cases), or as in some edge cases (or anything else) are currently being worked on? I sure hope it's not the first one.

clason commented:
It's the first one unless proven otherwise. (Whence no progress in two years.)

Andrew15-5 commented:
> I guess your problem occurs, because you are calling the function from within visual mode.

But I `call` `feedkeys('gvy')` function also from within visual mode. I execute them exactly the same way (commands and function).

One note is that in all of my commands I remove the `'<,'>` prefix (I'm pretty sure I don't need it and manual commands work this way).

MultisampledNight commented:
@Andrew15-5 Are you on any more sync chat? I think this would fit better on Matrix or the like.

<details>
<summary>Anyway, this is what I use to get the current visual selection in my `init.vim`. You can just check for non-`v` beforehand and execute `gv` first in case you also want to handle the last one if not in visual mode.</summary>

```vim
function SelectionToClipboard()
	if mode() == "v"
		let selection_start = getcurpos()[1:]
		silent normal! o
		let selection_end = getcurpos()[1:]

		silent normal! "*y

		call cursor(selection_start)
		silent normal! v
		call cursor(selection_end)
		silent normal! o
	endif
endfunction
```

</details>

Andrew15-5 commented:
I checked it. It obviously doesn't return the selected text, but even when I add this functionality, it doesn't give me what I want (same as with my function).

I personally need the text extraction and returning in the same function (like mine).

---

> Are you on any more sync chat? I think this would fit better on Matrix or the like.

I'm currently not (well, if we are not talking about Telegram). But of course I can create one more account. But I have no intent to have a big conversation. Unless there is something to talk about/discuss, we should keep things as they are.

clason commented:
Then we already talked too much

nyngwang commented:
Is it possible to integrate this into `nvim_create_user_command`? So that I can get visual-selected region when calling a command upon selection like `'<,'>:MyCommand`.

justinmk commented:
# Status

- Closing this because of lack of activity. 
- To continue this effort, create a new PR that addresses the following:
     - `vim.region` already exists, it should be enhanced to (better) handle visual selections. E.g. currently it can't handle block selection.
    - If getting text from the result of `vim.region()` is difficult, maybe we need a new function like  `vim.region_text({region})` that converts a `vim.region()` result to a string.

See also: https://github.com/neovim/neovim/pull/13896/files#r779819674
